### PR TITLE
Deprecate PathSearch, remove internal use

### LINKED ISSUE #17120: Deprecate and remove `PathSearch` class
I am looking at our `PathSearch` class and I absolutely do not like this thing.

I think such functionality is best deferred to either `std::filesystem` [1] or the boost filesystem class [2]. There is absolutely no reason to reimplement such functionality again in deal.II. On top of that, the special `"MESH"` functionality has not been updated in years. It does not find half of our currently supported mesh files.

While at it, I think we should remove the use of this class from `GridIn::read()`. These functions should not search but simply trying to open the supplied filepath (or stream object) and fail if unsuccessful.

Quick vote and opinions?

[1] https://en.cppreference.com/w/cpp/filesystem
[2] https://www.boost.org/doc/libs/1_34_0/libs/filesystem/doc/index.htm

### PR DESCRIPTION
**Pull request by tamiko**
This pull request deprecates the class and removes the internal use in the GridIn class.

Closes #17120

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]

**Review by marcfehling — APPROVED** [review]

File: include/deal.II/base/path_search.h [hunk 1]
```diff
@@ -75,11 +75,11 @@ DEAL_II_NAMESPACE_OPEN
  * characters are not added automatically (allowing you to do some real file
  * name editing).
  *
- * @todo Add support for environment variables like in kpathsea.
+ * @deprecated Use std::filesystem or boost::filesystem facilities instead.
```
Comment (bangerth) [inline include/deal.II/base/path_search.h:78]: We don't ship `boost::filesystem`'s source files, only the headers. Let's just not recommend it.

**Review by bangerth — COMMENTED** [review]
OK if you remove the one mention of `boost::filesystem`. People should use the standard facilities.

**Review by drwells — APPROVED** [review]

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
This fails a bunch of tests:
[projectroot.tests.grid.grid_in_msh_version_2debug.serial.grid/grid_in_msh_version_2.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-17276/6/testReport/junit/projectroot.tests.grid.grid_in_msh_version_2debug/serial/grid_grid_in_msh_version_2_debug/)
[projectroot.tests.grid.grid_in_msh_v4debug.serial.grid/grid_in_msh_v4.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-17276/6/testReport/junit/projectroot.tests.grid.grid_in_msh_v4debug/serial/grid_grid_in_msh_v4_debug/)
[projectroot.tests.grid.grid_in_unv_3ddebug.serial.grid/grid_in_unv_3d.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-17276/6/testReport/junit/projectroot.tests.grid.grid_in_unv_3ddebug/serial/grid_grid_in_unv_3d_debug/)
[projectroot.tests.grid.grid_in_mshdebug.serial.grid/grid_in_msh.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-17276/6/testReport/junit/projectroot.tests.grid.grid_in_mshdebug/serial/grid_grid_in_msh_debug/)
[projectroot.tests.grid.grid_in_msh_03_v4debug.serial.grid/grid_in_msh_03_v4.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-17276/6/testReport/junit/projectroot.tests.grid.grid_in_msh_03_v4debug/serial/grid_grid_in_msh_03_v4_debug/)
[projectroot.tests.grid.grid_in_msh_version_1debug.serial.grid/grid_in_msh_version_1.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-17276/6/testReport/junit/projectroot.tests.grid.grid_in_msh_version_1debug/serial/grid_grid_in_msh_version_1_debug/)
[projectroot.tests.grid.grid_in_unv_2ddebug.serial.grid/grid_in_unv_2d.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-17276/6/testReport/junit/projectroot.tests.grid.grid_in_unv_2ddebug/serial/grid_grid_in_unv_2d_debug/)
[projectroot.tests.grid.grid_in_unv_3d_02debug.serial.grid/grid_in_unv_3d_02.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-17276/6/testReport/junit/projectroot.tests.grid.grid_in_unv_3d_02debug/serial/grid_grid_in_unv_3d_02_debug/)
[projectroot.tests.grid.grid_indebug.serial.grid/grid_in.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-17276/6/testReport/junit/projectroot.tests.grid.grid_indebug/serial/grid_grid_in_debug/)
[projectroot.tests.grid.grid_in_msh_03debug.serial.grid/grid_in_msh_03.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-17276/6/testReport/junit/projectroot.tests.grid.grid_in_msh_03debug/serial/grid_grid_in_msh_03_debug/)

**Comment by bangerth** [comment]
Example:
```
An error occurred in line <4142> of file </jenkins/workspace/dealii-serial_PR-17276/source/grid/grid_in.cc> in function
    void dealii::GridIn<dim, spacedim>::read(const string&, dealii::GridIn<dim, spacedim>::Format) [with int dim = 2; int spacedim = 2; std::string = std::__cxx11::basic_string<char>]
The violated condition was: 
    std::filesystem::exists(filename)
Additional information: 
    Could not open file
    /jenkins/workspace/dealii-serial_PR-17276/tests/grid/grid_in_msh_version_2/hole81.
```
This fails because the code looks like this (note that we don't specify a file extension):
```
  check_file<2>(std::string(SOURCE_DIR "/grid_in_msh_version_2/hole81"),
                GridIn<2>::msh);
```

**Comment by tamiko** [comment]
Alright, I have checked and, if necessary, updated all tests that contain the string `GridIn` and everything under `grid/`.

**Review by tjhei — APPROVED** [review]

**Pull request merged**